### PR TITLE
API Checker for sock-shop Application to track requests on endpoints

### PR DESCRIPTION
**Pull request by oumkale**
**What this PR does / why we need it**:

API checker send requests to ```/user, /catalogue, /details, /cards, /addresses, /basket``` endpoints.
Logs can be seen. We can see easily chaos injection change on endpoints easily using this api-checker.

**Special notes for your reviewer**:
With successfully case :+1: 
```
[Status]: API Checker has been started
[Status]: ResponseCode: 200  User successfully logged in
[Info]: Unique ID for user is: qcB5EZVSax5Gc1372Qzqnv__tK4dHXoW
[Info]: Sending request to front-end
[Status]: ResponseCode: 200  FrontEnd is accessible
[Info]: Adding new cutomer 
[Status]: ResponseCode: 200  Customer added successfully with user Name: test_user_cd74905e-0cd7-4d64-9fcb-38069d9f86de
[Status]: ResponseCode: 200  Catalogue get request successfully send
[Info]: Catalogue Item: {'id': '808a2de1-1aaa-4c25-a9b9-6612e8f29a38', 'name': 'Crossed', 'description': 'A mature sock, crossed, with an air of nonchalance.', 'imageUrl': ['/catalogue/images/cross_1.jpeg', '/catalogue/images/cross_2.jpeg'], 'price': 17.32, 'count': 738, 'tag': ['red', 'formal', 'blue', 'action']}
[Info]: Adding card details for purchase
[Status]: ResponseCode: 200  Card details has been successfully added
[Info]: Adding Address for user
[Status]: ResponseCode: 200  Address has been added successfully
[Status]: RespondeCode: 200  Card ID: {'longNum': '5953580604169678', 'expires': '08/19', 'ccv': '678', 'id': '57a98d98e4b00679b4a830ae', '_links': {'card': {'href': 'http://user/cards/57a98d98e4b00679b4a830ae'}, 'self': {'href': 'http://user/cards/57a98d98e4b00679b4a830ae'}}}
[Info]: Getting Address
[Status]: ResponseCode: 200  Address has been retrieved successfully
[Info]: Getting item details
[Status]: ResponseCode: 200  Item details retrieved successfully

```
With user-db Failure case :-1: 
```
[Error]: Unable to send requests to user-db, Server encountered a condition that is preventing request.
[Error]: HTTPConnectionPool(host='front-end.sock-shop.svc.cluster.local', port=80): Read timed out. (read timeout=1)
[Error]: Unable to send requests to used-db, Server encountered a condition that is preventing request.
[Error]: HTTPConnectionPool(host='front-end.sock-shop.svc.cluster.local', port=80): Read timed out. (read timeout=1)
[Status]: ResponseCode: 200  Catalogue get request successfully send
[Info]: Catalogue Item: {'id': 'd3588630-ad8e-49df-bbd7-3167f7efb246', 'name': 'YouTube.sock', 'description': "We were not paid to sell this sock. It's just a bit geeky.", 'imageUrl': ['/catalogue/images/youtube_1.jpeg', '/catalogue/images/youtube_2.jpeg'], 'price': 10.99, 'count': 801, 'tag': ['formal', 'geek']}
[Error]: Unable to send requests to user-db, Server encountered a condition that is preventing request.
[Error]: HTTPConnectionPool(host='front-end.sock-shop.svc.cluster.local', port=80): Read timed out. (read timeout=1)
[Error]: Unable to send requests to used-db, Server encountered a condition that is preventing request.
[Error]: HTTPConnectionPool(host='front-end.sock-shop.svc.cluster.local', port=80): Read timed out. (read timeout=1)
[Status]: ResponseCode: 200  Catalogue get request successfully send
[Info]: Catalogue Item: {'id': 'a0a4f044-b040-410d-8ead-4de0446aec7e', 'name': 'Nerd leg', 'description': 'For all those leg lovers out there. A perfect example of a swivel chair trained calf. Meticulously trained on a diet of sitting and Pina Coladas. Phwarr...', 'imageUrl': ['/catalogue/images/bit_of_leg_1.jpeg', '/catalogue/images/bit_of_leg_2.jpeg'], 'price': 7.99, 'count': 115, 'tag': ['skin', 'blue']}
[Error]: Unable to send requests to user-db, Server encountered a condition that is preventing request.
[Error]: HTTPConnectionPool(host='front-end.sock-shop.svc.cluster.local', port=80): Read timed out. (read timeout=1)
[Error]: Unable to send requests to used-db, Server encountered a condition that is preventing request.
[Error]: HTTPConnectionPool(host='front-end.sock-shop.svc.cluster.local', port=80): Read timed out. (read timeout=1)

```

### CODE REVIEW

File: custom/git-app-deployer/api-checker/main.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+            #basket
+            requests.get(url  +  "/basket.html")
+
+        except Exception as exc:
+            print("[Error] : Unable to send requests, Server encountered a condition that is preventing request.")
+            print('[Error] : {err}'.format(err=exc))
+        
+        time.sleep(8)       
+
+print("[Status] : API Checker has been started")
```
Comment (ispeakc0de) [inline custom/git-app-deployer/api-checker/main.py:102]:
```suggestion
print("[Status]: API Checker has been started")
```

File: custom/git-app-deployer/api-checker/main.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+my_function()
```
Comment (ispeakc0de) [inline custom/git-app-deployer/api-checker/main.py:103]: can we use better name?

File: custom/git-app-deployer/api-checker/main.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+""" URL of the api endpoints of application to send and recieve requests """
+url = os.environ['URL']
+
+def my_function():
+    while True:
+        try:
+            base64_str = base64.encodebytes(('%s:%s' % ("user","password")).encode()).decode().strip()
+            
+            #login
+            print("[Info] : User is login")
```
Comment (ispeakc0de) [inline custom/git-app-deployer/api-checker/main.py:22]: we can update it

File: custom/git-app-deployer/api-checker/main.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            
+            login = requests.get(url  +  "/login", headers={"Authorization":"Basic %s" % base64_str})
+            with login as response:
+                if response.status_code == 200:
+                    print("[Status] : " , response.status_code, " User logged in successfully. ")
```
Comment (ispeakc0de) [inline custom/git-app-deployer/api-checker/main.py:27]:
```suggestion
                    print("[Status]: " , response.status_code, " User logged successfully")
```

File: custom/git-app-deployer/api-checker/main.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+                    
+            cust_id = login.cookies["logged_in"]
```
Comment (ispeakc0de) [inline custom/git-app-deployer/api-checker/main.py:29]: camel case?

File: custom/git-app-deployer/api-checker/main.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+            cookies = {'logged_id': cust_id}
+            print("[Info] : Unique ID for user is:", cust_id)
```
Comment (ispeakc0de) [inline custom/git-app-deployer/api-checker/main.py:31]:
```suggestion
            print("[Info]: Unique ID for user is: ", cust_id)
```

File: custom/git-app-deployer/api-checker/main.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+            
+            print("[Info] : Sending request to front-end")
```
Comment (ispeakc0de) [inline custom/git-app-deployer/api-checker/main.py:33]:
```suggestion
            print("[Info]: Sending request to front-end")
```

File: custom/git-app-deployer/api-checker/main.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+            #front-end
+            frontend = requests.get(url  +  "/")
+            with frontend as response:
+                if response.status_code == 200:
+                    print("[Status] : " , response.status_code, " Front-End is accessible. ")
```
Comment (ispeakc0de) [inline custom/git-app-deployer/api-checker/main.py:38]:
```suggestion
                    print("[Status]:  ResponseCode: " , response.status_code, " FrontEnd is accessible")
```

File: custom/git-app-deployer/api-checker/main.py [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+            #catalogue
+            catalogue = requests.get(url  +  "/catalogue")
+            with catalogue as response:
+                if response.status_code == 200:
+                    category_item = choice(catalogue.json())
+                    item_id = category_item["id"]
+                    print("[Status] : " , response.status_code, " Catalogue get request successfully send. ")
```
Comment (ispeakc0de) [inline custom/git-app-deployer/api-checker/main.py:46]:
```suggestion
                    print("[Status]:  ResponseCode: " , response.status_code, " Catalogue get request successfully send")
```

File: custom/git-app-deployer/api-checker/main.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+                    print("[Info] : Catalogue Items :", category_item)
```
Comment (ispeakc0de) [inline custom/git-app-deployer/api-checker/main.py:47]:
```suggestion
                    print("[Info]: Catalogue Item: ", category_item)
```

File: custom/git-app-deployer/api-checker/main.py [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+            
+            #register 
+            """Counter is creating uniqe ID for customer registration"""
+            global counter
+            counter += 1
+            username = "test_user_" + str(uuid.uuid4())
+            password = "test_password"
+
+            print("[Info] : Adding cutomer ")
```
Comment (ispeakc0de) [inline custom/git-app-deployer/api-checker/main.py:56]:
```suggestion
            print("[Info]: Adding new cutomer ")
```

File: custom/git-app-deployer/api-checker/main.py [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+            with requests.post(url  +  "/register", json={"username": username, "password": password}) as response:
+                if response.status_code == 200:
+                    print("[Status] : " , response.status_code, " Cutomer added successfully send with User Name:",  username)
```
Comment (ispeakc0de) [inline custom/git-app-deployer/api-checker/main.py:59]:
```suggestion
                    print("[Status]: ResponseCode:" , response.status_code, " Cutomer added successfully with user Name:",  username)
```

File: custom/git-app-deployer/api-checker/main.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+            
+            #cards
+            print("[Info] : Adding card details for purchase.")
```
Comment (ispeakc0de) [inline custom/git-app-deployer/api-checker/main.py:62]:
```suggestion
            print("[Info]: Adding card details for purchase")
```

File: custom/git-app-deployer/api-checker/main.py [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+            with requests.post(url  +"/cards", json={"longNum": "5429804235432", "expires": "04/16", "ccv": "432", "userId": cust_id},headers={"Authorization":"Basic %s" % base64_str}) as response:
+                if response.status_code == 200:
+                    print("[Status] : " , response.status_code, " Card details has been successfully added. ")
```
Comment (ispeakc0de) [inline custom/git-app-deployer/api-checker/main.py:65]:
```suggestion
                    print("[Status]: ResponseCode:" , response.status_code, "Card details has been successfully added")
```

File: custom/git-app-deployer/api-checker/main.py [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+            
+            #addresses
+            print("[Info] : Adding Address details for user.")
```
Comment (ispeakc0de) [inline custom/git-app-deployer/api-checker/main.py:68]:
```suggestion
            print("[Info]: Adding Address for user")
```

File: custom/git-app-deployer/api-checker/main.py [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+            with requests.post(url  +  "/addresses", json={"street": "my road", "number": "3", "country": "UK", "city": "London"},headers={"Authorization":"Basic %s" % base64_str}) as response:
+                if response.status_code == 200:
+                    print("[Status] : " , response.status_code, " Address details has been successfully added with. ")
```
Comment (ispeakc0de) [inline custom/git-app-deployer/api-checker/main.py:71]:
```suggestion
                    print("[Status]: ResponseCode:" , response.status_code, "Address has been added successfully")
```

File: custom/git-app-deployer/api-checker/main.py [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+            
+            #cards
+            cards = requests.get(url  +"/cards")
+            with cards as response:
+                if response.status_code == 200:
+                    cardDetails = cards.json()
+                    card_id = choice(cardDetails["_embedded"]["card"])
+                    print("[Status] : " , response.status_code, " Card details : ",card_id)
```
Comment (ispeakc0de) [inline custom/git-app-deployer/api-checker/main.py:79]:
```suggestion
                    print("[Status]: RespondeCode:" , response.status_code, "Card ID:",card_id)
```

File: custom/git-app-deployer/api-checker/main.py [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+            
+            #addresses
+            with requests.get(url  +  "/addresses") as response:
+                if response.status_code == 200:
+                    print("[Status] : " , response.status_code, " Address details has been successfully recieved.")
```
Comment (ispeakc0de) [inline custom/git-app-deployer/api-checker/main.py:84]:
```suggestion
                    print("[Status]: ResponseCode:" , response.status_code, " Address has been retrieved successfully")
```

File: custom/git-app-deployer/api-checker/main.py [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+            
+            #details
+            print("[Info] : Get item details")
```
Comment (ispeakc0de) [inline custom/git-app-deployer/api-checker/main.py:87]:
```suggestion
            print("[Info]: Getting item details")
```

File: custom/git-app-deployer/api-checker/main.py [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+            requests.get(url  +  "/category.html")
+            requests.get(url  +  "/detail.html?id={}".format(item_id))
+            with requests.get(url  +  "/detail.html?id={}".format(item_id)) as response:
+                if response.status_code == 200:
+                    print("[Status] : " , response.status_code, " Dtails of item successfully recieved.")
```
Comment (ispeakc0de) [inline custom/git-app-deployer/api-checker/main.py:92]:
```suggestion
                    print("[Status]: ResponseCode:" , response.status_code, " Item details successfully recieved")
```

File: custom/git-app-deployer/api-checker/main.py [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+            #basket
+            requests.get(url  +  "/basket.html")
+
+        except Exception as exc:
+            print("[Error] : Unable to send requests, Server encountered a condition that is preventing request.")
```
Comment (ispeakc0de) [inline custom/git-app-deployer/api-checker/main.py:97]:
```suggestion
            print("[Error]: Unable to send request, Server encountered a condition that is preventing request")
```

File: custom/git-app-deployer/api-checker/main.py [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+            print('[Error] : {err}'.format(err=exc))
```
Comment (ispeakc0de) [inline custom/git-app-deployer/api-checker/main.py:98]:
```suggestion
            print('[Error]: {err}'.format(err=exc))
```

File: custom/git-app-deployer/api-checker/main.py [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+            requests.get(url  +  "/basket.html")
+
+        except Exception as exc:
+            print("[Error]: Unable to send request, Server encountered a condition that is preventing request")
+            print('[Error]: {err}'.format(err=exc))
+        
+        time.sleep(8)       
+
+print("[Status]: API Checker has been started")
+my_function()
```
Comment (ispeakc0de) [inline custom/git-app-deployer/api-checker/main.py:103]: different Name

**Review by ispeakc0de — DISMISSED** [review]

**Review by ispeakc0de — APPROVED** [review]

**Review by chandankumar4 — APPROVED** [review]

**Pull request merged**